R/proj.R

```diff
@@ -54,40 +68,66 @@ proj_get <- function(quiet = FALSE) {
 #'   project-signalling infrastructure, such as initialising a Git repo or
 #'   adding a DESCRIPTION file.
 #' @export
-proj_set <- function(path = ".", force = FALSE, quiet = FALSE) {
-  if (!is.null(path)) {
-    check_is_dir(path)
-  }
-  path <- proj_path_prep(path)
-
-  if (!force) {
-    new_project <- proj_path_prep(proj_find(path))
-    if (is.null(new_project)) {
-      stop_glue(
-        "Path {value(path)} does not appear to be inside a project or package."
-      )
-    }
-    path <- new_project
+proj_set <- function(path = ".", force = FALSE) {
+  if (is.null(path) || force) {
+    path <- proj_path_prep(path)
+    proj_string <- if (is.null(path)) code("NULL") else value(path)
+    done("Setting active project to {proj_string}")
+    return(proj_set_(path))
   }
 
-  proj_set_(path, quiet = quiet)
+  check_is_dir(path)
+  new_project <- proj_find(path)
+  if (is.null(new_project)) {
+    stop_glue(
+      "Path {value(path)} does not appear to be inside a project or package."
+    )
+  }
+  proj_set(path = new_project, force = TRUE)
 }
 
-#' @describeIn proj_utils Builds a path within the active project. Thin wrapper
-#'   around [fs::path()].
+#' @describeIn proj_utils Builds a path within the active project returned by
+#'   `proj_get()`. Thin wrapper around [fs::path()].
 #' @inheritParams fs::path
 #' @export
 proj_path <- function(..., ext = "") {
   path_norm(path(proj_get(), ..., ext = ext))
 }
 
-proj_set_ <- function(path, quiet = FALSE) {
-  old <- proj$cur
-  proj$cur <- path
-  if (!quiet) {
-    done("Setting active project to {value(proj$cur)}")
-  }
-  invisible(old)
+#' @describeIn proj_utils Runs code with a temporary active project. It is an
+#'   example of the `with_*()` functions in [withr](http://withr.r-lib.org).
+#' @param code Code to run with temporary active project.
+#' @param quiet Whether to suppress user-facing messages, while operating in the
+#'   temporary active project.
+#' @export
+with_project <- function(path = ".",
+                         code,
+                         force = FALSE,
+                         quiet = getOption("usethis.quiet", default = FALSE)) {
+  old_quiet <- options(usethis.quiet = quiet)
+  old_proj  <- proj_set(path = path, force = force)
+
+  on.exit(proj_set(path = old_proj, force = TRUE))
```

Comment:
I maybe would just combine these together into one `on.exit()` call.

```r
on.exit({
  proj_set(path = old_proj, force = TRUE)
  options(old_quiet)
})
```

Reply:
Yeah, that also makes the `with_` and `local_` code more parallel. Done.